mlflow/models/signature.py

```diff
@@ -0,0 +1,449 @@
+                input_example.shape))
+        input_example = pd.DataFrame(input_example)
+    elif not isinstance(input_example, pd.DataFrame):
+        raise TypeError("Unexpected type of input_example. Expected one of "
+                        "(pandas.DataFrame, numpy.ndarray, dict, list), got {}".format(
+                          type(input_example)))
+
+    example_filename = "input_dataframe_example.json"
+    res = input_example.to_dict(orient="split")
+    # Do not include row index
+    del res["index"]
+    if all(input_example.columns == range(len(input_example.columns))):
+        # No need to write default column index out
+        del res["columns"]
+
+    with open(os.path.join(path, example_filename), "w") as f:
+        json.dump(res, f, cls=NumpyEncoder)
+    return example_filename
+
+
+def _base64decode(x):
+    return base64.decodebytes(x.encode("ascii"))
+
+
+def dataframe_from_json(path_or_str, schema: Schema = None,
+                        pandas_orient: str = "split") -> pd.DataFrame:
+    """
+    Parse json into pandas.DataFrame. User can pass schema to ensure correct type parsing and to
+    make any necessary conversions (e.g. string -> binary for binary columns).
+
+    :param path_or_str: Path to a json file or a json string.
+    :param schema: Mlflow schema used when parsing the data.
+    :param pandas_orient: pandas data frame convention used to store the data.
+    :return: pandas.DataFrame.
+    """
+    if schema is not None:
+        dtypes = dict(zip(schema.column_names(), schema.column_types()))
+        df = pd.read_json(path_or_str, orient=pandas_orient, dtype=dtypes)[schema.column_names()]
+        binary_cols = [i for i, x in enumerate(schema.column_types()) if x == DataType.binary]
+
+        for i in binary_cols:
+            col = df.columns[i]
+            df[col] = np.array(df[col].map(_base64decode), dtype=np.bytes_)
+            return df
+    else:
+        return pd.read_json(path_or_str, orient=pandas_orient, dtype=False)
+
+
+def _map_numpy_dtype(col: np.ndarray) -> DataType:
+    if not isinstance(col, np.ndarray):
+        raise TypeError("Expected numpy.ndarray, got '{}'.".format(type(col)))
+    if len(col.shape) > 1:
+        raise MlflowException("Expected 1d array, got array with shape {}".format(col.shape))
+    if col.dtype.kind == "b":
+        return DataType.boolean
+    elif col.dtype.kind == "i" or col.dtype.kind == "u":
+        if col.dtype.itemsize < 4 or col.dtype.kind == "i" and col.dtype.itemsize == 4:
+            return DataType.integer
+        elif col.dtype.itemsize < 8 or col.dtype.kind == "i" and col.dtype.itemsize == 8:
+            return DataType.long
+    elif col.dtype.kind == "f":
+        if col.dtype.itemsize <= 4:
+            return DataType.float
+        elif col.dtype.itemsize <= 8:
+            return DataType.double
+    elif col.dtype.kind == "U":
+        return DataType.string
+    elif col.dtype.kind == "S":
```

Comment:
Do we have unit tests for all these cases? Would be good to just create a variety of ndarrays and confirm they all work, so I don't have to read the numpy documentation.